### PR TITLE
Gamepad API implementation

### LINKED ISSUE #10977: Implement the Gamepad API
Request to implement the awesome [Gamepad API](https://w3c.github.io/gamepad/). The API introduces `navigator.getGamepads()`; interfaces (`Gamepad`, `GamepadButton`, `GamepadEvent`); and [`gamepadconnected` and `gamepaddisconnected` events](https://w3c.github.io/gamepad/#event-gamepadconnected).

Here are a few helpful links:
- https://w3c.github.io/gamepad/
- https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API/Using_the_Gamepad_API
- https://www.smashingmagazine.com/2015/11/gamepad-api-in-web-games/
- http://www.html5rocks.com/en/tutorials/doodles/gamepad/

### PR DESCRIPTION
**Pull request by MortimerGoro**
Gamepad API implementation. Tested with HTC Vive and Daydream controllers ;)

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #10977

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because : current gamepad tests are manual (https://github.com/w3c/web-platform-tests/tree/master/gamepad). There is a  open issue about the best way to test WebVR/Gamepad without real devices https://github.com/w3c/webvr/issues/187. We'll work on the testing suite & mock devices/data on a separate issue.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16260)

### CODE REVIEW

File: components/script/dom/gamepadbuttonlist.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    #[allow(unrooted_must_root)]
+    fn new_inherited(list: Vec<JS<GamepadButton>>) -> GamepadButtonList {
+        GamepadButtonList {
+            reflector_: Reflector::new(),
+            list: list
+        }
+    }
+
+    #[allow(unrooted_must_root)]
+    pub fn new(global: &GlobalScope, list: Vec<Root<GamepadButton>>) -> Root<GamepadButtonList> {
```
Comment (nox) [inline components/script/dom/gamepadbuttonlist.rs]: Pass a `&[&GamepadButton]`.

File: components/script/dom/gamepadbuttonlist.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        reflect_dom_object(box GamepadButtonList::new_inherited(list.iter().map(|r| JS::from_ref(&**r)).collect()),
```
Comment (nox) [inline components/script/dom/gamepadbuttonlist.rs]: Pass the `&[&GamepadButton]` value and do the collect in `new_inherited`.

File: components/script/dom/gamepadbuttonlist.rs (hunk 1 above)
Comment (nox) [inline components/script/dom/gamepadbuttonlist.rs]: Is this method even needed? I can't see it called anywhere. If it's not, just remove it.

File: components/script/dom/gamepadbuttonlist.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    pub fn new_from_vr(global: &GlobalScope, buttons: &[WebVRGamepadButton]) -> Root<GamepadButtonList> {
+        let mut list = Vec::new();
+        for btn in buttons {
+            list.push(GamepadButton::new(&global, btn.pressed, btn.touched));
+        }
+
+        GamepadButtonList::new(&global, list)
+    }
+
+    pub fn sync_vr(&self, vr_buttons: &[WebVRGamepadButton]) {
```
Comment (nox) [inline components/script/dom/gamepadbuttonlist.rs]: You seem to have named similar methods `update_from_vr`, please stay consistent.

File: components/script/dom/gamepadbuttonlist.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+impl GamepadButtonListMethods for GamepadButtonList {
+    // https://www.w3.org/TR/gamepad/#dom-gamepad-buttons
+    fn Length(&self) -> u32 {
+        self.list.len() as u32
+    }
+
+    // https://www.w3.org/TR/gamepad/#dom-gamepad-buttons
+    fn Item(&self, index: u32) -> Option<Root<GamepadButton>> {
+        if (index as usize) < self.list.len() {
```
Comment (nox) [inline components/script/dom/gamepadbuttonlist.rs]:
```rust
self.list.get(index as usize).map(|button| Root::from_ref(&*button))
```

File: components/script/dom/gamepadevent.rs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    pub fn new_with_type(global: &GlobalScope, event_type: GamepadEventType, gamepad: &Gamepad)
+                         -> Root<GamepadEvent> {
+        let name = match event_type {
+            GamepadEventType::Connected => "gamepadconnected",
+            GamepadEventType::Disconnected => "gamepaddisconnected"
+        };
+
+        GamepadEvent::new(&global,
+                          Atom::from(DOMString::from(name)),
```
Comment (nox) [inline components/script/dom/gamepadevent.rs]: Just `name.into()` or `Atom::from(name)` is enough here.

File: components/script/dom/gamepadevent.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                          false,
+                          false,
+                          &gamepad)
+    }
+
+    pub fn Constructor(window: &Window,
```
Comment (nox) [inline components/script/dom/gamepadevent.rs]: Spec link.

File: components/script/dom/gamepadlist.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+// https://www.w3.org/TR/gamepad/
+#[dom_struct]
+pub struct GamepadList {
+    reflector_: Reflector,
+    list: DOMRefCell<Vec<JS<Gamepad>>>
+}
+
+impl GamepadList {
+    #[allow(unrooted_must_root)]
+    fn new_inherited(list: Vec<JS<Gamepad>>) -> GamepadList {
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]: Pass a `&[&Gamepad]`.

File: components/script/dom/gamepadlist.rs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+// https://www.w3.org/TR/gamepad/
+#[dom_struct]
+pub struct GamepadList {
+    reflector_: Reflector,
+    list: DOMRefCell<Vec<JS<Gamepad>>>
+}
+
+impl GamepadList {
+    #[allow(unrooted_must_root)]
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]: Why is this needed?

File: components/script/dom/gamepadlist.rs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        reflect_dom_object(box GamepadList::new_inherited(list.iter().map(|r| JS::from_ref(&**r)).collect()),
+                           global,
+                           GamepadListBinding::Wrap)
+    }
+
+    pub fn add(&self, gamepad: &Gamepad) {
+        self.list.borrow_mut().push(JS::from_ref(&*gamepad));
+    }
+
+    pub fn sync(&self, gamepads: &[Root<Gamepad>]) {
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]: Pass a `&[&Gamepad]`.

File: components/script/dom/gamepadlist.rs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        for gamepad in gamepads {
+            if self.list.borrow().iter().find(|g| g.gamepad_id() == gamepad.gamepad_id()).is_none() {
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]: Use `Iterator::any`.

File: components/script/dom/gamepadlist.rs (hunk 9 above)
Comment (nox) [inline components/script/dom/gamepadlist.rs]: Why is this named `sync`?

File: components/script/dom/gamepadlist.rs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+impl GamepadListMethods for GamepadList {
+    // https://www.w3.org/TR/gamepad/#dom-navigator-getgamepads
+    fn Length(&self) -> u32 {
+        self.list.borrow().len() as u32
+    }
+
+    // https://www.w3.org/TR/gamepad/#dom-navigator-getgamepads
+    fn Item(&self, index: u32) -> Option<Root<Gamepad>> {
+        if (index as usize) < self.list.borrow().len() {
+            Some(Root::from_ref(&*(self.list.borrow()[index as usize])))
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]:
```rust
self.list.borrow().get(index as usize).map(|gamepad| Root::from_ref(&*gamepad))
```

File: components/script/dom/gamepadlist.rs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    #[allow(unrooted_must_root)]
+    fn new_inherited(list: Vec<JS<Gamepad>>) -> GamepadList {
+        GamepadList {
+            reflector_: Reflector::new(),
+            list: DOMRefCell::new(list)
+        }
+    }
+
+    #[allow(unrooted_must_root)]
+    pub fn new(global: &GlobalScope, list: Vec<Root<Gamepad>>) -> Root<GamepadList> {
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]: Pass a `&[&Gamepad]`.

File: components/script/dom/navigator.rs [hunk 13]
```diff
@@ -128,15 +131,26 @@ impl NavigatorMethods for Navigator {
...
+    }
     // https://w3c.github.io/permissions/#navigator-and-workernavigator-extension
     fn Permissions(&self) -> Root<Permissions> {
         self.permissions.or_init(|| Permissions::new(&self.global()))
     }
 }
 
 impl Navigator {
-     pub fn handle_webvr_event(&self, event: WebVREventMsg) {
+     pub fn handle_webvr_events(&self, events: Vec<WebVREventMsg>) {
```
Comment (nox) [inline components/script/dom/navigator.rs]: We shouldn't need to go through the `Navigator` object just to handle `WebVR` events, should we? Please file a follow-up issue about this.

File: components/script/dom/vr.rs [hunk 14]
```diff
@@ -20,20 +22,21 @@ use ipc_channel::ipc;
 use ipc_channel::ipc::IpcSender;
 use script_traits::WebVREventMsg;
 use std::rc::Rc;
-use webvr_traits::WebVRMsg;
-use webvr_traits::webvr;
+use webvr_traits::*;
```
Comment (nox) [inline components/script/dom/vr.rs]: No glob import, please. Use what you need and nothing more.

File: components/script/dom/vr.rs [hunk 15]
```diff
@@ -128,34 +131,112 @@ impl VR {
...
-                self.notify_event(&display, &event);
-            },
-            &webvr::VRDisplayEvent::Disconnect(id) => {
-                if let Some(display) = self.find_display(id) {
-                    display.handle_webvr_event(&event);
-                    self.notify_event(&display, &event);
-                }
+        match event {
+            WebVREventMsg::DisplayEvent(event) => {
+                match &event {
```
Comment (nox) [inline components/script/dom/vr.rs]: Remove `&`.

File: components/script/dom/vr.rs [hunk 16]
```diff
@@ -128,34 +131,112 @@ impl VR {
...
-            &webvr::VRDisplayEvent::Activate(ref display, _) |
-            &webvr::VRDisplayEvent::Deactivate(ref display, _) |
-            &webvr::VRDisplayEvent::Blur(ref display) |
-            &webvr::VRDisplayEvent::Focus(ref display) |
-            &webvr::VRDisplayEvent::PresentChange(ref display, _) |
-            &webvr::VRDisplayEvent::Change(ref display) => {
-                let display = self.sync_display(&display);
-                display.handle_webvr_event(&event);
+            WebVREventMsg::GamepadEvent(event) => {
+                match &event {
```
Comment (nox) [inline components/script/dom/vr.rs]: Remove `&`.

File: components/script/dom/vr.rs [hunk 17]
```diff
@@ -128,34 +131,112 @@ impl VR {
...
-            },
-            &webvr::VRDisplayEvent::Disconnect(id) => {
-                if let Some(display) = self.find_display(id) {
-                    display.handle_webvr_event(&event);
-                    self.notify_event(&display, &event);
-                }
+        match event {
+            WebVREventMsg::DisplayEvent(event) => {
+                match &event {
+                    &WebVRDisplayEvent::Connect(ref display) => {
```
Comment (nox) [inline components/script/dom/vr.rs]: The change from `webvr::VRDisplayEvent` to `WebVRDisplayEvent` should live in its own commit to make reviewing easier.

File: components/script/dom/vr.rs [hunk 18]
```diff
@@ -128,34 +131,112 @@ impl VR {
...
             },
-            &webvr::VRDisplayEvent::Activate(ref display, _) |
-            &webvr::VRDisplayEvent::Deactivate(ref display, _) |
-            &webvr::VRDisplayEvent::Blur(ref display) |
-            &webvr::VRDisplayEvent::Focus(ref display) |
-            &webvr::VRDisplayEvent::PresentChange(ref display, _) |
-            &webvr::VRDisplayEvent::Change(ref display) => {
-                let display = self.sync_display(&display);
-                display.handle_webvr_event(&event);
+            WebVREventMsg::GamepadEvent(event) => {
```
Comment (nox) [inline components/script/dom/vr.rs]: Please introduce two new private methods `handle_display_event` and `handle_gamepad_event` and call them from here.

File: components/script/dom/vr.rs [hunk 19]
```diff
@@ -128,34 +131,112 @@ impl VR {
...
+                    &WebVRGamepadEvent::Connect(ref data, ref state) => {
+                        if let Some(gamepad) = self.find_gamepad(state.gamepad_id) {
+                            gamepad.update_from_vr(&state);
+                        } else {
+                            // new gamepad
+                            self.sync_gamepad(Some(data.clone()), state);
+                        }
+                    },
+                    &WebVRGamepadEvent::Disconnect(id) => {
+                        if let Some(gamepad) = self.find_gamepad(id) {
```
Comment (nox) [inline components/script/dom/vr.rs]: Can this event actually be received with no corresponding gamepad?

File: components/script/dom/vr.rs [hunk 20]
```diff
@@ -128,34 +131,112 @@ impl VR {
...
+// Gamepad
+impl VR {
+    fn find_gamepad(&self, gamepad_id: u64) -> Option<Root<Gamepad>> {
+        self.gamepads.borrow()
+                     .iter()
+                     .find(|g| g.gamepad_id() == gamepad_id)
+                     .map(|g| Root::from_ref(&**g))
+    }
+
+    fn sync_gamepad(&self, data: Option<WebVRGamepadData>, state: &WebVRGamepadState) {
```
Comment (nox) [inline components/script/dom/vr.rs]: Take a `Option<&WebVRGamepadData>` here, this avoids a cloning operation when handling `WebVRGamepadEvent::Connect` above.

File: components/script_traits/lib.rs [hunk 21]
```diff
@@ -280,8 +280,8 @@ pub enum ConstellationControlMsg {
     ReportCSSError(PipelineId, String, usize, usize, String),
     /// Reload the given page.
     Reload(PipelineId),
-    /// Notifies the script thread of a WebVR device event
-    WebVREvent(PipelineId, WebVREventMsg)
+    /// Notifies the script thread of WebVR events.
+    WebVREvent(PipelineId, Vec<WebVREventMsg>)
```
Comment (nox) [inline components/script_traits/lib.rs]: Rename this `WebVREvents`.

File: components/script_traits/lib.rs [hunk 22]
```diff
@@ -751,16 +751,18 @@ pub enum ConstellationMsg {
     LogEntry(Option<FrameId>, Option<String>, LogEntry),
     /// Set the WebVR thread channel.
     SetWebVRThread(IpcSender<WebVRMsg>),
-    /// Dispatch a WebVR event to the subscribed script threads.
-    WebVREvent(Vec<PipelineId>, WebVREventMsg),
+    /// Dispatch WebVR events to the subscribed script threads.
+    WebVREvent(Vec<PipelineId>, Vec<WebVREventMsg>),
```
Comment (nox) [inline components/script_traits/lib.rs]: Rename this `WebVREvents`.

File: components/webvr/webvr_thread.rs [hunk 23]
```diff
@@ -230,16 +250,20 @@ impl WebVRThread {
...
         let pipeline_ids: Vec<PipelineId> = self.contexts.iter().map(|c| *c).collect();
-        for event in events {
-            let event = WebVREventMsg::DisplayEvent(event);
-            self.constellation_chan.send(ConstellationMsg::WebVREvent(pipeline_ids.clone(), event)).unwrap();
-        }
+        let events = events.drain(0..).map(|ev| {
+            match ev {
+                VREvent::Display(ev) => WebVREventMsg::DisplayEvent(ev),
+                VREvent::Gamepad(ev) => WebVREventMsg::GamepadEvent(ev),
+            }
```
Comment (nox) [inline components/webvr/webvr_thread.rs]: Why do we even have a `WebVREventMsg` type if `VREvent` is exactly the same?

File: python/tidy/servo_tidy/tidy.py [hunk 24]
```diff
@@ -74,6 +74,7 @@
     "//url.spec.whatwg.org",
     "//xhr.spec.whatwg.org",
     "//w3c.github.io",
+    "//www.w3.org",
```
Comment (nox) [inline python/tidy/servo_tidy/tidy.py]: No, change the spec links.

**Review by nox — CHANGES_REQUESTED** [review]
First review pass.

File: components/script/dom/vr.rs (hunk 20 above)
Comment (MortimerGoro) [inline components/script/dom/vr.rs]: Changed to use move semantics, so the clone above is not required either

File: components/script/dom/gamepad.rs [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+use std::cell::Cell;
+use std::ptr;
+use webvr_traits::{WebVRGamepadData, WebVRGamepadHand, WebVRGamepadState};
+
+#[dom_struct]
+pub struct Gamepad {
+    reflector_: Reflector,
+    gamepad_id: Cell<u64>,
+    id: DOMRefCell<String>,
+    index: Cell<u32>,
```
Comment (MortimerGoro) [inline components/script/dom/gamepad.rs]: Now it's mutable to ensure that the index is correct if more Gamepad providers are added in the future.

File: components/script/dom/vr.rs (hunk 19 above)
Comment (MortimerGoro) [inline components/script/dom/vr.rs]: Yes, it can happen in new "tabs"/JSContexts when the gamepads where already connected in the previous pages.

File: components/webvr/webvr_thread.rs (hunk 23 above)
Comment (MortimerGoro) [inline components/webvr/webvr_thread.rs]: I did that to have the possibility to add some internal events in the future. But it's not needed now, so I'll change that only if required in the future.

File: components/script/dom/gamepad.rs (hunk 25 above)
Comment (nox) [inline components/script/dom/gamepad.rs]: If more gamepad providers are added, why would the index of this particular gamepad change?

Comment (MortimerGoro) [inline components/script/dom/gamepad.rs]: I'm talking about not WebVR related gamepad providers. The index is unkown until it is added to the global gampepad array here: https://github.com/servo/servo/pull/16260/commits/42f8a8110f55b8f504baaa3835ee63578e3d8091#diff-4ca88ffda625c3517cee6640aac60b99R140 The index must match the final position in the array.

Comment (nox) [inline components/script/dom/gamepad.rs]: I guess I don't understand why this field is needed then. Could you document it?

Comment (nox) [inline components/script/dom/gamepad.rs]: Oh I understand now. Could you find a different design for this? I'm pretty sure we don't even need the index at all for now, it would matter only if you had hundreds of gamepads. No need to keep the index in the list in mind, just linearly search for it in the `GamepadList` object.

Comment (MortimerGoro) [inline components/script/dom/gamepad.rs]: In order to do the dynamic search I would need to add a reference to the Gamepadlist to a Gamepad object (Gamepads are unaware of the container) which seems worse than a plain i32 automatically updated by the gamepadlist. Another way would be to get the gamepadlist from window->navigator->gamepadlist.  Not too fond on both ideas, but I prefer the second solution. What do you think?

**Comment by nox** [comment]
What is the point of the GamepadList object if it doesn't hold onto the gamepads? What hold onto them currently?

**Comment by MortimerGoro** [comment]
The GamepadList object has 2 main purposes:
- It's used as a fast & GC friendly container to share the gamepad list between rust and javascript, without creating/updating JS arrays every frame.
- Allows to implement the IndexGetter which will be used to hide gamepads according to privacy rules. The gamepad draft allows things like [null, [object Gamepad]]

In the prev comment I meant that Gamepad objects don't have a "parent" equivalent reference now, So I need to add the reference or go throught window.navigator to perform the linear search for the index property.

**Comment by nox** [comment]
Sorry, I still don't understand why you need this index. I feel like this lacks a lot of documentation.

**Comment by nox** [comment]
OOOH, disregard everything I just said, only now I realise the index is actually exposed to the Web and it is a concept defined in the spec. I guess storing the index isn't a bad idea then. Will review the rest of the changes you made.

**Comment by MortimerGoro** [comment]
Ok, I'll store it as i32 field then. I was confused when reading the spec too, it feels weird to expose the  index of an external inside the gamepad object, but it's defined that way: https://www.w3.org/TR/gamepad/#dom-gamepad-index

File: components/script/dom/gamepad.rs [hunk 26]
```diff
@@ -0,0 +1,208 @@
...
+    pose: Option<JS<VRPose>>,
+    #[ignore_heap_size_of = "Defined in rust-webvr"]
+    hand: WebVRGamepadHand,
+    display_id: u32
+}
+
+unsafe_no_jsmanaged_fields!(WebVRGamepadHand);
+
+impl Gamepad {
+    pub fn new_inherited(gamepad_id: u32,
```
Comment (nox) [inline components/script/dom/gamepad.rs]: Doesn't need to be public.

File: components/script/dom/gamepad.rs [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+    mapping_type: String,
+    axes: Heap<*mut JSObject>,
+    buttons: JS<GamepadButtonList>,
+    pose: Option<JS<VRPose>>,
+    #[ignore_heap_size_of = "Defined in rust-webvr"]
+    hand: WebVRGamepadHand,
+    display_id: u32
+}
+
+unsafe_no_jsmanaged_fields!(WebVRGamepadHand);
```
Comment (nox) [inline components/script/dom/gamepad.rs]: Put that in the `trace` module, not here in the middle of nowhere.

File: components/script/dom/gamepadbuttonlist.rs [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+use dom::bindings::codegen::Bindings::GamepadButtonListBinding;
+use dom::bindings::codegen::Bindings::GamepadButtonListBinding::GamepadButtonListMethods;
+use dom::bindings::js::{JS, Root};
+use dom::bindings::reflector::{Reflector, reflect_dom_object};
+use dom::gamepadbutton::GamepadButton;
+use dom::globalscope::GlobalScope;
+use dom_struct::dom_struct;
+use webvr_traits::WebVRGamepadButton;
+
+// https://www.w3.org/TR/GamepadButton/
```
Comment (nox) [inline components/script/dom/gamepadbuttonlist.rs]: This spec link is wrong.

File: components/script/dom/gamepadbuttonlist.rs [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+#[dom_struct]
+pub struct GamepadButtonList {
+    reflector_: Reflector,
+    list: Vec<JS<GamepadButton>>
+}
+
+impl GamepadButtonList {
+    #[allow(unrooted_must_root)]
+    fn new_inherited(list: Vec<JS<GamepadButton>>) -> GamepadButtonList {
```
Comment (nox) [inline components/script/dom/gamepadbuttonlist.rs]: Pass `&[&GamepadButton]` here (see the `r` method on `[JS<T>]`).

File: components/script/dom/gamepadlist.rs [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    pub fn new(global: &GlobalScope, list: &[&Gamepad]) -> Root<GamepadList> {
+        reflect_dom_object(box GamepadList::new_inherited(list),
+                           global,
+                           GamepadListBinding::Wrap)
+    }
+
+    fn add(&self, gamepad: &Gamepad) {
+        self.list.borrow_mut().push(JS::from_ref(&*gamepad));
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]: What happens if the gamepad was already in the list?

File: components/script/dom/gamepadlist.rs [hunk 31]
```diff
@@ -0,0 +1,69 @@
...
+        // Ensure that the gamepad has the correct index
+        gamepad.update_index(self.list.borrow().len() as i32 - 1);
+    }
+
+    pub fn add_if_not_exists(&self, gamepads: &[Root<Gamepad>]) {
+        for gamepad in gamepads {
+            if self.list.borrow().iter().any(|g| g.gamepad_id() == gamepad.gamepad_id()) {
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]: This will add the gamepad only if it is already in the list.

File: components/script/dom/gamepadlist.rs [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+impl GamepadListMethods for GamepadList {
+    // https://w3c.github.io/gamepad/#dom-navigator-getgamepads
+    fn Length(&self) -> u32 {
+        self.list.borrow().len() as u32
+    }
+
+    // https://w3c.github.io/gamepad/#dom-navigator-getgamepads
+    fn Item(&self, index: u32) -> Option<Root<Gamepad>> {
+        if (index as usize) < self.list.borrow().len() {
+            Some(Root::from_ref(&*(self.list.borrow()[index as usize])))
```
Comment (nox) [inline components/script/dom/gamepadlist.rs]: Do the same thing as in `GamepadButtonList::Item` here.

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by nox** [comment]
Cc @larsbergstrom or anyone else with a VR kit that want to manually test this branch.

File: components/script/dom/gamepadlist.rs (hunk 30 above)
Comment (MortimerGoro) [inline components/script/dom/gamepadlist.rs]: This method was only called internally by add_if_not_exists. I have removed it to avoid confusions with the add_if_not_exists method.

File: components/script/dom/gamepadbuttonlist.rs (hunk 29 above)
Comment (MortimerGoro) [inline components/script/dom/gamepadbuttonlist.rs]: Thanks, the r() method is very useful for this ;)

**Comment by nox** [comment]
This needs a squash, and I would like @larsbergstrom to take a quick look. LGTM otherwise.

**Review by nox — APPROVED** [review]

File: .gitignore [hunk 33]
```diff
@@ -27,3 +27,5 @@ Servo.app
 
 # VSCode
 .vscode
+
+*.iml
```
Comment (nox) [inline .gitignore]: What is this file extension btw?

Comment (MortimerGoro) [inline .gitignore]: It's a file automatically created by Android Studio (I use it to test on Android).

I'll remove it from here because that change belongs more to https://github.com/servo/servo/pull/15773, which is about to be merged

**Comment by MortimerGoro** [comment]
Squash ready ;)

**Comment by larsbergstrom** [comment]
@bors-servo r=nox,larsbergstrom

I've tested the gamekit stuff and can confirm controllers work :-)

**Comment by danShumway** [comment]
Did this ever make its way into master?

I'm still getting ``navigator.getGamepads is not a function``.  Looking around, I don't see a build flag mentioned anywhere. Any chance there's something obvious I'm missing?

**Comment by danShumway** [comment]
Thanks! Just for future reference so I don't need to bother people as much, I assume [this](https://github.com/servo/servo/blob/master/resources/prefs.json) is the list of existing preferences?
